### ISSUE TITLE
3rd line of holes 

### ISSUE THREAD

**Issue by TEDSTANDAGE** [CCDevNelson/Snap-Updates#58]
I accidentally posted this to SNaP-Dev I think, so I closed it there and re-created it here. Sorry for that.

Is SNaP exporting a 3rd line drilling of holes when hanging or shelving is set back? The drawings currently do not show a 3rd line of holes when there is a setback, but perhaps the XML reflects the 3rd line of holes? If not, we need to add this, and the drawing should reflect the line of holes as well.

The following shows a 16" deep section with the hang and dust shelf set back 4", but with no 3rd line of holes showing up:

![image](https://user-images.githubusercontent.com/61025894/81444009-45ddb380-912b-11ea-8724-7c11bfa64d15.png)

**Comment by TEDSTANDAGE** [comment]
Teddy, please have Ryan prioritize this item as it goes with the other drilling improvements we have been doing. Thanks, Ted

**Comment by kategodfrey** [comment]
@TedClassy1209 Can we have the option of having that third line drilled in construction options? @JBTattersall was wanting that third line for a shelf stack. I'm not sure of what other instances that we would need it....

**Comment by TedClassy1209** [comment]
@kategodfrey we have a fix here: https://github.com/CCDevNelson/SNaP-Dev/issues/186. I don't think we have tested this yet though. Correct me if I am wrong @ClassyTeddy, but I think the idea was for it to automatically drill the 3rd line of holes when the user has selected a setback for a rod or for shelves (or when the rod is automatically setback to 12" in 16" or deeper sections per https://github.com/CCDevNelson/Snap-Updates/issues/78). 

However, to the point of having it available in construction options @JBTattersall @kategodfrey, there are times when designers request 3 lines of holes even when all the parts ordered match the default (whether for possible future hangs in 16"+ deep sections, or otherwise), so it would definitely make sense to have the option to drill the 3rd line of holes in the construction options (and to have it reflected here when the 3rd line shows up per https://github.com/CCDevNelson/SNaP-Dev/issues/186). 

Another note for @ClassyTeddy when creating this option: Default should be that the 3rd line would match the front lines of holes in a 12" deep section, with the option to have it be in a custom location.

**Comment by ClassyTeddy** [comment]
I thought this was tested and it did work. It should be automatically drilling a third line of holes when something has been setback.

If you want to put it in my column I can try and add the option for the third line of holes to be selected by the user, but what happens in a scenario where the user selects a third line of holes with a custom location, and the 3rd line of hole is placed automatically because of a setback being selected? Should it be forced to pick the automatic holes, or the holes set with a custom location by the user.

I personally feel like it should be automatic or manual, but not both.

**Comment by TedClassy1209** [comment]
We may have tested it looking at code, etc., but I don't think we have physically drilled a 3rd line of holes yet out in the shop to make sure it is accurately being reflected in the XML/label's barcodes/etc.

I think it should remain automatic as well, but we need the option to have a 3rd line of holes even when there is no setback of rods or shelves currently existing in the opening (like I said, for a future rod or other configuration). So, maybe we can't have the automatic additional lines of holes from rod/shelf setbacks be reflected in the construction options, but we do need the option to add a 3rd line of holes in the construction options that perhaps comes with a popup that reminds the user that 3rd lines of holes are already automatically created from setbacks they have made to the rods/shelves?

**Comment by ClassyTeddy** [comment]
Unfortunately, I cannot do a popup inside of a prompt menu, because the prompt menu is itself, a popup. So users will just have to be aware of it, or I can add some text to the prompt menu saying that.

**Comment by TedClassy1209** [comment]
Okay yeah, then maybe if they select a 3rd line of holes option in the construction options then a line of text can appear below it reminding the user about the automatic 3rd line of holes (would prefer not to add more additional lines to the standard construction options prompt menu, it is already really crowded, so if we can have it hidden until they select the 3rd line option that would be best).

**Comment by TedCC1209** [comment]
I do not believe a third line of holes is being drilled--I cannot see it in Closet Machining Setup or in the XML.  The is the same job I emailed to @CCDevNelson and others earlier this week to test this issue, two 16" deep partitions with a rod set back 4":

![no 3 lines0](https://user-images.githubusercontent.com/61025894/107692675-5dbe0700-6c6a-11eb-8aac-b77c46f41ab5.png)
![no 3 lines](https://user-images.githubusercontent.com/61025894/107692594-3f580b80-6c6a-11eb-90f0-8a11e961b9ea.png)
![no 3 lines2](https://user-images.githubusercontent.com/61025894/107692610-4252fc00-6c6a-11eb-9bdb-0a094f587ab4.png)

Perhaps I am doing this wrong?

Looking at the XML on that same job I couldn't see a hole location for the 3rd line of holes either, only the front line and the back line for a 16" deep partition (14.54"+/- from the front and 1.68" from the front, where there should also be a line of holes 5.68" from the front to account for the 4" rod setback). Here is the XML file:

[snap_job.zip](https://github.com/CCDevNelson/Snap-Updates/files/5968192/snap_job.zip)

Referencing the following issue where I think a 3rd line of holes was previously addressed: https://github.com/CCDevNelson/SNaP-Dev/issues/186

**Comment by RyanMDev** [comment]
The 3rd line of holes depends on the "Add Rod Setback" prompt:
```
            set_back = insert.get_prompt("Add Rod Setback")
            left_panel, right_panel = self.get_adj_panels(insert_bp)

            if set_back and set_back.value() > macp.dim_to_front_system_hole:
```
https://github.com/CCDevNelson/SNaP-Dev/blob/a2b07b4b5c0deba91959d4f6882cdd177f875cba/SNaP/2.78/scripts/addons/snap_db/library_scripts/Library-Classy_Closets/mv_closet_machining.py#L468

Previously this prompt was available on the hanging rods prompt menu but has since been removed:

![hanging_rod_prompts_122](https://user-images.githubusercontent.com/6529544/108896561-f1c49280-75c9-11eb-906c-d492286bd61e.png)

However, this prompt can still be accessed in the sidebar with the hanging rod insert selected:

![hanging_rod_prompts_sidebar](https://user-images.githubusercontent.com/6529544/108897365-df972400-75ca-11eb-93db-514c73cc512d.png)

**Comment by TylerStandage** [comment]
@TedCC1209 See the above comment from Ryan and let us know if we can close this issue.

**Comment by ClassyTeddy** [comment]
This issue is ready for testing on the `Fixing-Third-Line-Of-Holes` issue

**Comment by TedCC1209** [comment]
I am now seeing 3 lines of holes, thanks Teddy. The third line of holes looks to be in the incorrect place, however, for this 16" deep partition with a rod setback 4" to be at a standard 12" partition location:

![3 lines of holes](https://user-images.githubusercontent.com/61025894/109555398-c1319c80-7a92-11eb-8091-c7669a133440.png)

The "Y" coordinates in the XML are as follows for the three lines in the above example are coming out as follows:

1 - ~1.68"
2 - ~3.14"
3 - ~14.54"

I'm a little confused as to what this is measuring from--if from the back of the partition, then the third line of holes (line "2", above) should be located at ~10.54":

- the back line of holes should be ~1.68" from the back of the partition
- the front line of holes should be ~1.46" from the front of the partition, or 14.54" from the back of a 16" deep partition, and 
- the middle or 3rd line of holes should be located exactly 4" back from the front line of holes, i.e. ~5.46" from the front of the 16" deep partition, or 10.54" from the back of the 16" deep partition).

**Comment by ClassyTeddy** [comment]
Ready for testing on the `Fixing-Third-Line-Of-Holes` branch.

**Comment by TedCC1209** [comment]
I am not seeing any third line of holes showing up now? Are you seeing them @ClassyTeddy ? If so, maybe I am doing it wrong.

**Comment by ClassyTeddy** [comment]
I am seeing them. Make sure that you place a new object. Save, export the room, open the room again, and then they should change.

**Comment by TedCC1209** [comment]
> I am seeing them. Make sure that you place a new object. Save, export the room, open the room again, and then they should change.

Yep. In Debug Mode, created new project, created new room, inserted a one hanging opening assembly with two partitions (16" deep), put in double hang that automatically sets the rod back 4", saved the room, exported the room's XML, reopened the room and I see the two partitions in the Closet Machining Setup but they only have the front and back lines of holes.

**Comment by ClassyTeddy** [comment]
Alright, so I got it to show up on a partition that is less than 16 inches deep with a setback rod, but it doesn't show up for any partition larger than 16 inches. Let me go ahead and fix that.

**Comment by ClassyTeddy** [comment]
Alright, should be good to go on the `Fixing-Third-Line-of-Holes` branch.

**Comment by TedCC1209** [comment]
Okay it is showing up now, thanks. Note that 99% of the time we don't set back rods in sections less than 16", so the functionality is most important at 16+" deep.

I still don't think the middle line is showing up in the right place. In the same situation I have been testing (a 16" deep partition with the standard 4" rod setback) I now come up with the following locations for the holes:

1. 1.6899999419193044"
2. 5.456692584922397"
3. 14.543306859643785"

That middle line needs to be exactly 4" set back from the front line of holes (here, I would think it should be at 10.543306859643785" because I think these dimensions are from the back of the partition / i.e., the wall)

**Comment by ClassyTeddy** [comment]
All right, I'll take a look at the calculations again. I think it is placing them 4 inches back from the back holes instead of the front.
